[Tan Yu Qi] iP

yuqitanyq opened the pull request:
### Duke Chatbot
> “How much does it take to have a good idea? Zero. You don’t need time. You need room in your head.” -David Allen ([source](https://brandongaille.com/43-outstanding-david-allen-quotes/))

- Simple
- Text-based
- Responsive

All you need to do is,

1. Download it from this [link](https://github.com/yuqitanyq/ip/releases/tag/A-Jar)
2. Run the jar file
3. Add your tasks
4. ~~Mark~~ Check them done and undone accordingly
5. let it help you manage your tasks! 😁 

AND it's **FREE!**

Features:
- [x] List Tasks
- [ ] Check Tasks
- [ ] Find Tasks

If you are a Java Programmer, you can use it to practice Java too. Here's the `main` method:

```java

    public static void main(String[] args) {
        new Duke("data/duke.txt").run();
    }

```

Commit:
Add Gradle support

yuqitanyq committed:
Level 1. Greet, Echo, Exit

yuqitanyq committed:
Level 2. Add, List

yuqitanyq committed:
Level 3. Mark as Done

yuqitanyq committed:
Level 4. ToDos, Events, Deadlines (Created the child classes)

yuqitanyq committed:
Level 4. ToDos, Events, Deadlines

yuqitanyq committed:
Text UI Testing (Test using I/O redirection technique)

yuqitanyq committed:
Level 5. Handle Errors

yuqitanyq committed:
Level 6. Delete

yuqitanyq committed:
Refactor Code For Readability

The code base has become too long in Duke.java.

This makes it hard to read and add code into existing codebase.

Let's refactor and abstract out methods to improve readability.

yuqitanyq committed:
Change Name Conventions

Current naming conventions are not in imperative form.

Let's change them to imperative form.

yuqitanyq committed:
Level 7. Save

The current Duke does not enable a saving feature.

This makes the chatbot lose all its data once the program ends.

Let's enable a saving feature through file reader and file writer.

yuqitanyq committed:
Level 8. Dates and Times.

The chatbot still recognises date and time as strings.

We can do better than that. We can make sure the date gets recognised as date objects.

Let's parse the string date into the date object to create the date.

yuqitanyq committed:
Merge branch 'branch-Level-7'

yuqitanyq committed:
Merge branch 'branch-Level-8'

# Conflicts:
#	src/main/java/Duke.java
Resolve conflicts

yuqitanyq committed:
A-MoreOOP Refactor code to abstract out more classes

Current state has everything contained within Duke class.

This is not ideal for scaling of code.

Let's refactor the code to include the Ui class, Storage class, Parser class and Tasklist class.

yuqitanyq committed:
A-Packages Organise classes into suitable java packages

Currently there's no packaging for the classes.

This could lead to confusion as more classes are created.

Let's package the classes into various packages to organise them.

yuqitanyq committed:
A-JUnit Add several JUnit tests to ensure methods used are working well

Codebase does not have unit testing right now.

Unit Testing can be helpful in identifying bugs early.

Let's include JUnit tests to make sure code is tested regularly.

yuqitanyq committed:
Updated gitignore to include MANIFEST file

yuqitanyq committed:
A-JavaDoc Include JavaDocs for the codebase.

Current codebase does not have JavaDocs for documentation.

This makes it hard to understand code and debugging.

Let's include JavaDocs for all public methods and classes.

yuqitanyq committed:
A-CodingStandard Fix Code styling via coding standards.

Current code base has some issues regarding styling that violate coding standards.

Implementing the same coding standard will help increase readability of code.

Let's adjust the styling mistakes and rectify them according to coding standards.

yuqitanyq committed:
Level-9 Find tasks based on description

Current implementation does not allow users to search for a task.

Not being able to search in a long list might be problematic.

Let's create a search function.

yuqitanyq committed:
Merge branch 'branch-A-CodingStandard'

yuqitanyq committed:
Merge branch 'branch-Level-9'

yuqitanyq committed:
Merge branch 'branch-A-JavaDoc'

# Conflicts:
#	src/main/java/duke/functionality/TaskList.java

yuqitanyq committed:
Added some more JavaDocs

yuqitanyq committed:
Merge commit 'd8398594b7bc43da5eb865321c5a50cec4b3923d' into A-Gradle

yuqitanyq committed:
Merge branch 'branch-A-Gradle'

yuqitanyq committed:
Modify build.gradle file

AD-NAP commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+ * Deadline is a class which represents the deadline task.
+ */
+public class Deadline extends Task{
+    protected LocalDate by;
```
Perhaps the variable name here could be more descriptive? Maybe deadlineDate would describe it better?

AD-NAP commented on src/main/java/duke/tasks/Event.java:
```diff
@@ -0,0 +1,46 @@
+ * Event class is a class which represents the event task.
+ */
+public class Event extends Task{
+    protected LocalDate at;
```
Similar to Deadline, perhaps here could be eventDate? This allows easier add-ons in the future if you want to add eventTime for example.

AD-NAP commented on src/main/java/duke/functionality/TaskList.java:
```diff
@@ -0,0 +1,191 @@
+    /**
+     * Prints out the list of tasks in the tasklist.
+     */
+    public void listItems() {
```
Perhaps listTasks is a better name than listItems?

AD-NAP commented on src/main/java/duke/functionality/TaskList.java:
```diff
@@ -0,0 +1,191 @@
+        }
+    }
+
+    public void listItemsMatch() {
```
Similarly, perhaps this could be named as listMatchingTasks?

AD-NAP reviewed (commented):
Overall, I could not find much to point out under "other code quality guidelines" of the CS2103T Textbook. Great job on your codes. It's very clean and simple, yet understandable 😄 
Keep up the good work! Very impressive

yuqitanyq committed:
A-CheckStyle Use Checkstyle

Current implementation has no way to check if style is consistent.

This leads to inconsistency.

Let's include a way to check for styling.

yuqitanyq committed:
Revert "Modify build.gradle file"

This reverts commit d1f5dd6686c0f8d784e247576e8b0ea3fa68be93.

yuqitanyq committed:
Revert "Revert "Modify build.gradle file""

This reverts commit dfb4cc451ea09449bbd6810aff985b28d29bbd84.

yuqitanyq committed:
Merge branch 'branch-A-CheckStyle'

yuqitanyq committed:
Setting up files for GUI

yuqitanyq committed:
Level 10. GUI

Current state still uses a text based UI.

This is not user friendly.

Let's include a GUI into the chatbot to improve the user experience.

yuqitanyq committed:
Merge branch 'branch-Level-10'

glennljw commented on src/main/java/duke/functionality/TaskList.java:
```diff
@@ -0,0 +1,211 @@
+ * to allow users to track their tasks.
+ */
+public class TaskList {
+    private static final String LIST = "Here are the tasks in your list:\n";
+    private static final String MARK = "Nice! I've marked this task as done:\n  ";
+    private static final String UNMARK = "OK, I've marked this task as not done yet:\n  ";
+    private static final String ADDED = "Got it. I've added this task:\n  ";
+    private static final String REMOVED = "Noted. I've removed this task:\n";
+    private static final String ALREADY_MARKED = "Tasked has already been marked...";
+    private static final String ALREADY_UNMARKED = "Task has already been unmarked...";
+    private static final String ALREADY_DELETED = "Task has already been deleted...";
+    private static final String MATCH = "Here are the matching tasks in your list:\n";
```
Great use of constants!

glennljw commented on src/main/java/duke/tasks/Deadline.java:
```diff
@@ -0,0 +1,46 @@
+ * Deadline is a class which represents the deadline task.
+ */
+public class Deadline extends Task {
+    protected LocalDate by;
```
The name of the variable 'by' could perhaps be changed to something more easily identifiable for other readers? Maybe 'date' or 'deadlineDate'?

glennljw commented on src/main/java/duke/ui/UserInterface.java:
```diff
@@ -0,0 +1,42 @@
+    private static final String INTRO = "\nHello! I'm YQ, your assistant\n" + "What can I do for you?\n";
+    private static final String BYE = "\nBye. Hope to see you again soon!\n";
+
+    /**
+     * Prints the introduction message.
+     */
```
Perhaps the JavaDocs for this method could include @return?

glennljw commented on src/main/java/duke/ui/UserInterface.java:
```diff
@@ -0,0 +1,42 @@
+        return intro;
+    }
+
+    /**
+     * Prints the bye message.
+     */
```
Perhaps the JavaDocs for this method could include @return?

glennljw reviewed (commented):
Overall well done!

yuqitanyq committed:
Fix Frontend Bugs

yuqitanyq committed:
Fixed GUI Frontend Scrollpane

yuqitanyq committed:
Removed Logo

yuqitanyq committed:
Change Font Family For Send Button

yuqitanyq committed:
Fix Bugs Regarding Date Format When Parsing Date

yuqitanyq committed:
Updated All Javadocs comments.

yuqitanyq committed:
Add Changes Based On Reviews On Pull Requests.

yuqitanyq committed:
Update Storage and TaskList to include Assertions

Assertions are not used to check if tasks has indeed been added or deleted. Assertions are also not used to check if the directories has been created for the text file.

Using assertions asserts that tasks has been added for the various methods that add tasks to the tasklist. Also, it ensures that a new directory has indeed been created if the text file for storage does not exist.

Let's
Use Assertion to ensure that tasks have been added and deleted accordingly.
Use Assertions to ensure that the file directory has indeed been created.

yuqitanyq committed:
Improve Code Quality

Code Quality was not adhered to for Parser. Magic Numbers were being used for the indexing of substrings.

This makes code written hard to understand for readers of code base.

Let's,
Include constants for the different magic numbers so that the code is easier to understand.

yuqitanyq committed:
Improve Code Quality

Current codebase still contains redundant code.

This overcomplicates the code base for future development.

Let's
Remove the redundant code and clean up the code base.

yuqitanyq committed:
Merge pull request #1 from yuqitanyq/branch-A-Assertions

Update Storage and TaskList to include Assertions

yuqitanyq committed:
Merge pull request #2 from yuqitanyq/branch-A-CodeQuality

Improve Code Quality

yuqitanyq committed:
Create extension of add tasks feature to include detecting duplicates.

Current implementation allows duplicate tasks to be added.

This causes unnecessary trouble for the user, as adding duplicates does not add value.

Let's
Include a mechanism to check for duplicates based on description and dates and alert user if duplicates are added.

yuqitanyq committed:
Merge branch 'branch-Detect-Duplicates'

yuqitanyq committed:
Fix bugs on GUI

yuqitanyq committed:
Refactor code for cleaner directories.

yuqitanyq committed:
Fixed bug for deletion

yuqitanyq committed:
Changed background image and edited fonts

yuqitanyq committed:
Add a representative screenshot of the product to the docs folder

yuqitanyq committed:
Update the User Guide

Currently there's not user guide available.

This makes it hard for beginner users to use the chatbot.

Let's
Include the necessary features and examples on how to use them in the user guide.

yuqitanyq committed:
Set theme jekyll-theme-architect

yuqitanyq committed:
Fix bugs for dialogbox in GUI

Currently the dialogbox still shows eclipses after a certain number of tasks being added.

This makes it hard to read for the user.

Let's
Set the minimum height to USE_PREF_SIZE.

yuqitanyq committed:
Fix bugs for handling events and finding tasks

Current implementation does not prompt user if user forgets to input dates for events, and does not tell user that no tasks are found if there's no relevant results.

This does not give a clear course of action for the user.

Let's
Include exception handling for the case where events have missing dates.
Include a message for the user if there are no relevant tasks to be found.

yuqitanyq committed:
Fix Minor Bugs

Current implementation has bugs such as not throwing an error when a task does not exist in the list and invalid commands such as "todolist"

This confuses the user and would bring about bad user experience.

Let's
Throw an exception when the task does not exist.
Include a space in the command to be queried in the startsWith method.

yuqitanyq committed:
Changed the font back to System family

yuqitanyq committed:
Fix minor bugs

yuqitanyq committed:
Fix Checkstyle

yuqitanyq committed:
Set theme jekyll-theme-architect

yuqitanyq committed:
Improved documentation to include table of contents

yuqitanyq committed:
Merge branch 'branch-set-theme' and modify headers

yuqitanyq committed:
Changed header description

yuqitanyq committed:
Revert "Changed header description"

This reverts commit cb618ba7c3fd5e608be8819390db88fbcc3d89c8.

yuqitanyq committed:
Modify header description

yuqitanyq committed:
Edit spacing for header

yuqitanyq committed:
Modify instructions on how to launch application
